internal/controllers/machine/machine_controller_noderef.go

```diff
@@ -280,8 +253,33 @@ func (r *Reconciler) getNode(ctx context.Context, c client.Reader, providerID st
 func (r *Reconciler) patchNode(ctx context.Context, remoteClient client.Client, node *corev1.Node, newLabels, newAnnotations map[string]string, m *clusterv1.Machine) error {
 	newNode := node.DeepCopy()
 
-	// Adds the annotations CAPI sets on the node.
-	hasAnnotationChanges := annotations.AddAnnotations(newNode, newAnnotations)
+	// Adds the annotations from the Machine.
+	// NOTE: in order to handle deletion we are tracking the annotations set from the Machine in an annotation.
+	// At the next reconcile we are going to use this for deleting annotations previously set by the Machine, but
+	// not present anymore. Annotations not set from machines should be always preserved.
+	if newNode.Annotations == nil {
+		newNode.Annotations = make(map[string]string)
+	}
+	hasAnnotationChanges := false
+	annotationsFromPreviousReconcile := strings.Split(newNode.Annotations[clusterv1.AnnotationsFromMachineAnnotation], ",")
+	if len(annotationsFromPreviousReconcile) == 1 && annotationsFromPreviousReconcile[0] == "" {
+		annotationsFromPreviousReconcile = []string{}
+	}
+	// append well known names
+	annotationsFromPreviousReconcile = append(annotationsFromPreviousReconcile, CommonNodeAnnotations...)
+	// Adds the annotations CAPI sets on the node and computes the changes.
+	hasAnnotationChanges, annotationsFromCurrentReconcile := annotations.AddAnnotations(newNode, newAnnotations)
```

Comment:
annotationsFromCurrentReconcile is not the same as labelsFromCurrentReconcile below

labelsFromCurrentReconcile below also contains the labels that didn't change

I would suggest to simply stop using annotations.AddAnnotation and duplicating the log from below as much as possible (+/- necessary differences like for AnnotationsFromMachineAnnotation). This should be less error prone

The problem with the current implementation is that every annotation that we don't change in the current reconcile won't be included in AnnotationsFromMachineAnnotation anymore. This means we won't actually track any ownership after a few reconciles